### PR TITLE
ENH support any input length

### LINKED ISSUE #39: Input data shape for `DTCWT`
Now that `DTCWT` requires the input data to be a tensor of shape `(B, C, 2**J)`, where J is the number of wavelet scales, it can be inconvenient to apply to real data. However, the [dtcwt.np](https://github.com/rjw57/dtcwt/tree/master/dtcwt/numpy) package offers a less strict requirement for the input shape. Instead of `(B, C, 2**J)`, it only requires the input length to be even. To accommodate this, I made some modifications to achieve the same results.

### PR DESCRIPTION
**Pull request by xir4n**
resolves #39

### CODE REVIEW

File: tests/test_nn.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 @pytest.mark.parametrize("Q", [3, 4])
 @pytest.mark.parametrize("T", [8, 16])
 @pytest.mark.parametrize("padding_mode", ["symmetric", "zeros"])
-def test_shape(J, Q, T, padding_mode):
+def test_direct_shape(J, Q, T, padding_mode):
     B, C, N = 2, 3, 2**(J+4)
```
Comment (lostanlen) [inline tests/test_nn.py:11]: here i'd be curious to see a non-power of two shape for the unit test

**Review by lostanlen — CHANGES_REQUESTED** [review]

File: tests/test_nn.py (hunk 1 above)
Comment (xir4n) [reply tests/test_nn.py]: you're right, `DTCWT` supports both even and odd input length now

File: murenn/dtcwt/nn.py [hunk 2]
```diff
@@ -23,23 +23,25 @@ def __init__(self, *, J, Q, T, in_channels, padding_mode="symmetric"):
...
                 kernel_size=T,
                 bias=False,
                 groups=in_channels,
                 padding="same",
             )
-            torch.nn.init.normal_(conv1d.weight)
-            self.conv1d.append(conv1d)
+            torch.nn.init.normal_(conv1d_j.weight)
+            self.conv1d.append(conv1d_j)
+        # self.down = torch.nn.ModuleList(down)
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:43]: please remove rather than comment out

File: tests/test_nn.py [hunk 3]
```diff
@@ -18,7 +19,7 @@ def test_direct_shape(J, Q, T, padding_mode):
         padding_mode=padding_mode,
     )
     y = graph(x)
-    assert y.shape == (B, C, Q, J, 2**4)
+    assert y.shape[:4] == (B, C, Q, J)
```
Comment (lostanlen) [inline tests/test_nn.py:22]: ?

**Review by lostanlen — CHANGES_REQUESTED** [review]

File: tests/test_nn.py (hunk 3 above)
Comment (xir4n) [reply tests/test_nn.py:22]: Because of the padding effect, the length of y may be greater to $N/2^J$, I can't figure out a mathematical formula for the length of output, (which may depend on the filter length, the input data length, and also the decomposition level J), but y can still be passed to the second layer. Is it necessary to make it exactly equal to $N/2^J$?

File: murenn/dtcwt/nn.py (hunk 2 above)
Comment (xir4n) [reply murenn/dtcwt/nn.py]: sorry

File: tests/test_nn.py (hunk 3 above)
Comment (lostanlen) [reply tests/test_nn.py:22]: right now you seem to make it sound like the PyTorch array `y` may have more than four dimensions, and regardless of `N` and `J` i don't see how that can happen

Comment (lostanlen) [reply tests/test_nn.py:22]: i understand now, the fifth dimension is `N` and depends on the input signal

**Review by lostanlen — APPROVED** [review]

**Comment by lostanlen** [comment]
@xir4n many thanks for this. kindly merge at your convenience
